src/mlpack/methods/hdbscan/hdbscan_impl.hpp

```diff
@@ -0,0 +1,763 @@
+         typename MetricType,
+         template<typename TreeMetricType,
+                  typename TreeStatType,
+                  typename TreeMatType> class TreeType >
+HDBSCAN<NeighborSearch,
+        MetricType,
+        TreeType>::
+HDBSCAN(const size_t minPoints,
+        bool allowSingleCluster) :
+        minPoints(minPoints),
+        allowSingleCluster(allowSingleCluster)
+{
+  // Nothing to do.
+}
+
+/**
+ * Performs HDBSCAN clustering on the data,returning the list of cluster 
+ * assignments. The assigned clusters have values from 0 onwards. The noise 
+ * points are labelled as (total nuber of points + 1)
+ *
+ * Dcore -> distance of k-th nearest neighbor
+ *          here k is minPoints
+ * 1. Find the dcore of all the points.
+ * 2. Then append dcore to end of the input data (dataWithDcore)
+ * 3. Compute MST using the Dual Tree Borvuka Method.
+ * 4. Sort the obtained MST on basis of length of edges
+ *    in ascending order.
+ * 5. Covert MST to single likage tree .
+ *    Single linkage tree is a clustering method in which
+ *    we go on clustering the points until a single cluster
+ *    is formed.
+ * 6. Condense the obtained single linkage tree
+ *    to form condensed tree. Condensed tree contains only those
+ *    clusters as parent which have more then minPoints number of
+ *    points.
+ * 7. From this condense tree obtain labels of all the points.
+ */
+template<typename NeighborSearch,
+         typename MetricType,
+         template<typename TreeMetricType,
+                  typename TreeStatType,
+                  typename TreeMatType> class TreeType >
+template<typename MatType>
+void HDBSCAN<NeighborSearch,
+    MetricType,
+    TreeType>::
+Cluster(const MatType& data,
+        arma::Row<size_t>& assignments)
+{
+  // genearting a matrix with the values of dcore
+  arma::Mat<size_t> neighbors;
+  arma::mat distances;
+  neighborSearch.Train(data);
+  neighborSearch.Search(minPoints, neighbors, distances);
+  arma::mat dataWithDcore = arma::conv_to<arma::mat>::from(data);
+  dataWithDcore.resize(dataWithDcore.n_rows+1, dataWithDcore.n_cols);
+  dataWithDcore.row(dataWithDcore.n_rows-1) = distances.row(distances.n_rows-1);
+
+  // computing the mst of the given data and sort the obtained result
+  emst::DualTreeBoruvka<MetricType, arma::mat, TreeType>
+                       dtb(dataWithDcore, false, hMetric);
+  arma::mat resultsMST;
+  dtb.ComputeMST(resultsMST);
+
+  // create single linkage tree from the MST obtained.
+  arma::mat singleLinkageTree;
+  SingleLinkageTreeClustering(resultsMST, singleLinkageTree);
+
+  // Condense the obtained tree
+  arma::mat condensedTree;
+  CondenseTree(singleLinkageTree, condensedTree, minPoints);
+
+  // get clusters
+  arma::Mat<size_t> result;
+  GetClusters(condensedTree, result);
+
+  // save results obtained
+  assignments = result;
+}
+
+//  A class to help merge two clusters to form a
+// single linkage tree.
+class UnionForSingleLinkageClustering
```

Comment:
Is this the same as UnionFind from src/mlpack/methods/emst/union_find.hpp?  If so we can replace this with that.

Reply:
No, they are slightly different.
In emst/uinon_find.hpp when we do union, a point becomes parent of another point.
Instead what needs to be done is when we are doing union of two clusters, give the merged cluster a new index.
```
  void UnionOfTwoClusters(size_t m, size_t n)
  {
    // Each union will form a new cluster
    // with size as sum of its children
    sizeVector(nextLabel) = sizeVector(m) + sizeVector(n);
    parentVector(m) = nextLabel;
    parentVector(n) = nextLabel;
    nextLabel++;
  }
```
The nextLabel is making the difference.

Reply:
So, I am really a big fan of code reuse, so I think it is a good idea to try and use the existing `UnionFind` for this.  The difference that you are saying, basically, is that you are using `nextLabel` to label each cluster from `0` to the number of clusters.  But you already have a unique index---the index of the point.  I believe you can use this, and I would prefer that you do this: having two union find structures in the code is more maintenance for later and more tests for now that need to be written.

In any case... since the comments and the code are not really clear through `SingleLinkageTreeClustering()` and `CondenseTree()`, it is difficult for me to quickly understand the other changes that would be necessary.  Please consider what would need to be done to make this work though and let me know what you think.